File: Le_DARIEL/Services/AuthService.swift
//
//  AuthService.swift
//  Le_DARIEL
//
//  Created by FAEI971 on 23/07/2025.
//

import Foundation
import SwiftUI
import CryptoKit

class AuthService: ObservableObject {
    @Published var currentUser: User?
    @Published var isAuthenticated: Bool = false
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    
    private let userDefaults = UserDefaults.standard
    private let userKey = "currentUser"
    
    init() {
        loadCurrentUser()
    }
    
    // MARK: - Authentication
    
    func signUp(email: String, password: String, userType: UserType = .utilisateur1) async throws {
        isLoading = true
        defer { isLoading = false }
        
        // Vérifier si l'email existe déjà
        if await emailExists(email) {
            throw AuthError.emailAlreadyExists
        }
        
        // Créer le hash du mot de passe
        let passwordHash = hashPassword(password)
        
        // Créer l'utilisateur
        let user = User(
            email: email,
            passwordHash: passwordHash,
            userType: userType
        )
        
        // Sauvegarder l'utilisateur (dans un vrai projet, ce serait dans une base de données)
        try await saveUser(user)
        
        // Connecter l'utilisateur
        await MainActor.run {
            self.currentUser = user
            self.isAuthenticated = true
            saveCurrentUser()
        }
    }
    
    func signIn(email: String, password: String) async throws {
        isLoading = true
        defer { isLoading = false }
        
        // Récupérer l'utilisateur
        guard let user = await getUser(email: email) else {
            throw AuthError.userNotFound
        }
        
        // Vérifier le mot de passe
        let passwordHash = hashPassword(password)
        guard user.passwordHash == passwordHash else {
            throw AuthError.invalidPassword
        }
        
        // Mettre à jour la dernière connexion
        var updatedUser = user
        updatedUser.lastLogin = Date()
        try await saveUser(updatedUser)
        
        // Connecter l'utilisateur
        await MainActor.run {
            self.currentUser = updatedUser
            self.isAuthenticated = true
            saveCurrentUser()
        }
    }
    
    func signOut() {
        currentUser = nil
        isAuthenticated = false
        userDefaults.removeObject(forKey: userKey)
    }
    
    // MARK: - Password Management
    
    func resetPassword(email: String) async throws {
        // Dans un vrai projet, envoyer un email de réinitialisation
        // Pour l'instant, on simule
        guard await emailExists(email) else {
            throw AuthError.userNotFound
        }
        // TODO: Implémenter l'envoi d'email
    }
    
    // MARK: - User Management
    
    func upgradeToPremium() async throws {
        guard var user = currentUser else {
            throw AuthError.notAuthenticated
        }
        
        user.userType = .utilisateur2
        try await saveUser(user)
        
        await MainActor.run {
            self.currentUser = user
            saveCurrentUser()
        }
    }
    
    // MARK: - Private Helpers
    
    private func hashPassword(_ password: String) -> String {
        let data = Data(password.utf8)
        let hash = SHA256.hash(data: data)
        return hash.compactMap { String(format: "%02x", $0) }.joined()
    }
    
    private func emailExists(_ email: String) async -> Bool {
        // Dans un vrai projet, vérifier dans la base de données
        // Pour l'instant, on simule avec UserDefaults
        let users = loadAllUsers()
        return users.contains { $0.email.lowercased() == email.lowercased() }
    }
    
    private func getUser(email: String) async -> User? {
        let users = loadAllUsers()
        return users.first { $0.email.lowercased() == email.lowercased() }
    }
    
    private func saveUser(_ user: User) async throws {
        var users = loadAllUsers()
        if let index = users.firstIndex(where: { $0.id == user.id }) {
            users[index] = user
        } else {
            users.append(user)
        }
        saveAllUsers(users)
    }
    
    private func loadAllUsers() -> [User] {
        guard let data = userDefaults.data(forKey: "allUsers"),
              let users = try? JSONDecoder().decode([User].self, from: data) else {
            return []
        }
        return users
    }
    
    private func saveAllUsers(_ users: [User]) {
        if let data = try? JSONEncoder().encode(users) {
            userDefaults.set(data, forKey: "allUsers")
        }
    }
    
    private func loadCurrentUser() {
        guard let data = userDefaults.data(forKey: userKey),
              let user = try? JSONDecoder().decode(User.self, from: data) else {
            return
        }
        currentUser = user
        isAuthenticated = true
    }
    
    private func saveCurrentUser() {
        if let user = currentUser,
           let data = try? JSONEncoder().encode(user) {
            userDefaults.set(data, forKey: userKey)
        }
    }
}

// MARK: - Auth Errors

enum AuthError: LocalizedError {
    case emailAlreadyExists
    case userNotFound
    case invalidPassword
    case notAuthenticated
    case invalidEmail
    
    var errorDescription: String? {
        switch self {
        case .emailAlreadyExists:
            return "Cet email est déjà utilisé"
        case .userNotFound:
            return "Utilisateur non trouvé"
        case .invalidPassword:
            return "Mot de passe incorrect"
        case .notAuthenticated:
            return "Vous n'êtes pas connecté"
        case .invalidEmail:
            return "Email invalide"
        }
    }
}

File: lib/services/auth_service.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';
import 'package:flutter/foundation.dart';
import 'package:le_dariel/models/user.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';

class AuthService extends ChangeNotifier {
  User? _currentUser;
  bool _isAuthenticated = false;
  bool _isLoading = false;
  String? _errorMessage;

  User? get currentUser => _currentUser;
  bool get isAuthenticated => _isAuthenticated;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;

  AuthService() {
    _loadCurrentUser();
  }

  Future<void> signUp({
    required String email,
    required String password,
    UserType userType = UserType.utilisateur1,
  }) async {
    _isLoading = true;
    _errorMessage = null;

    try {
      if (await _emailExists(email)) {
        throw AuthError.emailAlreadyExists;
      }

      final passwordHash = _hashPassword(password);
      final user = User(
        id: const Uuid().v4(),
        email: email,
        passwordHash: passwordHash,
        userType: userType,
        createdAt: DateTime.now(),
      );

      await _saveUser(user);
      _currentUser = user;
      _isAuthenticated = true;
      await _saveCurrentUser();
      notifyListeners();
    } catch (e) {
      _errorMessage = e is AuthError ? e.message : 'Erreur lors de l\'inscription';
      rethrow;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  Future<void> signIn({
    required String email,
    required String password,
  }) async {
    _isLoading = true;
    _errorMessage = null;

    try {
      final user = await _getUser(email: email);
      if (user == null) {
        throw AuthError.userNotFound;
      }

      final passwordHash = _hashPassword(password);
      if (user.passwordHash != passwordHash) {
        throw AuthError.invalidPassword;
      }

      final updatedUser = User(
        id: user.id,
        email: user.email,
        passwordHash: user.passwordHash,
        userType: user.userType,
        createdAt: user.createdAt,
        lastLogin: DateTime.now(),
      );

      await _saveUser(updatedUser);
      _currentUser = updatedUser;
      _isAuthenticated = true;
      await _saveCurrentUser();
      notifyListeners();
    } catch (e) {
      _errorMessage = e is AuthError ? e.message : 'Erreur lors de la connexion';
      rethrow;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  Future<void> signOut() async {
    _currentUser = null;
    _isAuthenticated = false;
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('currentUser');
    notifyListeners();
  }

  Future<void> upgradeToPremium() async {
    if (_currentUser == null) {
      throw AuthError.notAuthenticated;
    }

    final updatedUser = User(
      id: _currentUser!.id,
      email: _currentUser!.email,
      passwordHash: _currentUser!.passwordHash,
      userType: UserType.utilisateur2,
      createdAt: _currentUser!.createdAt,
      lastLogin: _currentUser!.lastLogin,
    );

    await _saveUser(updatedUser);
    _currentUser = updatedUser;
    await _saveCurrentUser();
    notifyListeners();
  }

  String _hashPassword(String password) {
    final bytes = utf8.encode(password);
    final digest = sha256.convert(bytes);
    return digest.toString();
  }

  Future<bool> _emailExists(String email) async {
    final users = await _loadAllUsers();
    return users.any((u) => u.email.toLowerCase() == email.toLowerCase());
  }

  Future<User?> _getUser({required String email}) async {
    final users = await _loadAllUsers();
    try {
      return users.firstWhere(
        (u) => u.email.toLowerCase() == email.toLowerCase(),
      );
    } catch (e) {
      return null;
    }
  }

  Future<void> _saveUser(User user) async {
    final users = await _loadAllUsers();
    final index = users.indexWhere((u) => u.id == user.id);
    
    if (index != -1) {
      users[index] = user;
    } else {
      users.add(user);
    }
    
    await _saveAllUsers(users);
  }

  Future<List<User>> _loadAllUsers() async {
    final prefs = await SharedPreferences.getInstance();
    final data = prefs.getString('allUsers');
    if (data == null) return [];
    
    try {
      final List<dynamic> jsonList = json.decode(data);
      return jsonList.map((json) => User.fromJson(json)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<void> _saveAllUsers(List<User> users) async {
    final prefs = await SharedPreferences.getInstance();
    final jsonList = users.map((u) => u.toJson()).toList();
    await prefs.setString('allUsers', json.encode(jsonList));
  }

  Future<void> _loadCurrentUser() async {
    final prefs = await SharedPreferences.getInstance();
    final data = prefs.getString('currentUser');
    if (data == null) return;
    
    try {
      final json = jsonDecode(data);
      _currentUser = User.fromJson(json);
      _isAuthenticated = true;
      notifyListeners();
    } catch (e) {
      // Ignorer les erreurs de parsing
    }
  }

  Future<void> _saveCurrentUser() async {
    if (_currentUser == null) return;
    
    final prefs = await SharedPreferences.getInstance();
    final json = jsonEncode(_currentUser!.toJson());
    await prefs.setString('currentUser', json);
  }
}

class AuthError implements Exception {
  final String message;

  AuthError(this.message);

  static AuthError get emailAlreadyExists => 
      AuthError('Cet email est déjà utilisé');
  static AuthError get userNotFound => 
      AuthError('Utilisateur non trouvé');
  static AuthError get invalidPassword => 
      AuthError('Mot de passe incorrect');
  static AuthError get notAuthenticated => 
      AuthError('Vous n\'êtes pas connecté');
  static AuthError get invalidEmail => 
      AuthError('Email invalide');
}

